src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const clone = { ...state };
+
+  for (const object of actions) {
+    switch (object.type) {
+      case 'addProperties':
+        const { extraData } = object;
+
+        Object.assign(clone, extraData);
+        break;
+
+      case 'removeProperties':
+        const { keysToRemove } = object;
+
+        for (const key of keysToRemove) {
+          delete clone[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in clone) {
+          delete clone[key];
+        }
+    }
+    stateHistory.push({ ...clone });
```

Comment:
Good job on adhering to the DRY principle by pushing a copy of the state to `stateHistory` outside of the conditional logic.